gazebo_ros/src/gazebo_ros_properties.cpp

```diff
@@ -0,0 +1,448 @@
+            _res->geom_names.push_back(geom->GetName());
+          }
+        }
+      }
+    }
+
+    // get list of joints
+    _res->joint_names.clear();
+
+    gazebo::physics::Joint_V joints = model->GetJoints();
+    for (unsigned int i = 0; i < joints.size(); i++) {
+      _res->joint_names.push_back(joints[i]->GetName());
+    }
+
+    // get children model names
+    _res->child_model_names.clear();
+    for (unsigned int j = 0; j < model->GetChildCount(); j++) {
+      gazebo::physics::ModelPtr child_model =
+        boost::dynamic_pointer_cast<gazebo::physics::Model>(model->GetChild(j));
+      if (child_model) {
+        _res->child_model_names.push_back(child_model->GetName() );
+      }
+    }
+
+    // is model static
+    _res->is_static = model->IsStatic();
+
+    _res->success = true;
+    _res->status_message = "GetModelProperties: got properties";
+  }
+}
+
+void GazeboRosPropertiesPrivate::GetJointProperties(
+  gazebo_msgs::srv::GetJointProperties::Request::SharedPtr _req,
+  gazebo_msgs::srv::GetJointProperties::Response::SharedPtr _res)
+{
+  gazebo::physics::JointPtr joint;
+  for (unsigned int i = 0; i < world_->ModelCount(); ++i) {
+    joint = world_->ModelByIndex(i)->GetJoint(_req->joint_name);
+    if (joint) {
+      break;
+    }
+  }
+
+  if (!joint) {
+    _res->success = false;
+    _res->status_message = "GetJointProperties: joint not found";
+  } else {
+    /// @todo: FIXME
+    _res->type = _res->REVOLUTE;
+
+    _res->damping.clear();
+    _res->damping.push_back(joint->GetDamping(0));
+
+    _res->position.clear();
+    _res->position.push_back(joint->Position(0));
+
+    _res->rate.clear();  // use GetVelocity(i)
+    _res->rate.push_back(joint->GetVelocity(0));
+
+    _res->success = true;
+    _res->status_message = "GetJointProperties: got properties";
+  }
+}
+
+void GazeboRosPropertiesPrivate::GetLinkProperties(
+  gazebo_msgs::srv::GetLinkProperties::Request::SharedPtr _req,
+  gazebo_msgs::srv::GetLinkProperties::Response::SharedPtr _res)
+{
+  gazebo::physics::LinkPtr body =
+    boost::dynamic_pointer_cast<gazebo::physics::Link>(world_->EntityByName(_req->link_name));
+  if (!body) {
+    _res->success = false;
+    _res->status_message =
+      "GetLinkProperties: link not found, did you forget to scope the link by model name?";
+  } else {
+    /// @todo: validate
+    _res->gravity_mode = body->GetGravityMode();
+
+    gazebo::physics::InertialPtr inertia = body->GetInertial();
+
+    _res->mass = body->GetInertial()->Mass();
+
+    _res->ixx = inertia->IXX();
+    _res->iyy = inertia->IYY();
+    _res->izz = inertia->IZZ();
+    _res->ixy = inertia->IXY();
+    _res->ixz = inertia->IXZ();
+    _res->iyz = inertia->IYZ();
+
+    ignition::math::Vector3d com = body->GetInertial()->CoG();
+
+    _res->com.position.x = com.X();
+    _res->com.position.y = com.Y();
+    _res->com.position.z = com.Z();
+    _res->com.orientation.x = 0;  // @todo: gazebo do not support rotated inertia yet
```

Comment:
This is not true anymore, `auto com = body->GetInertial()->Pose()` should give you the translation and rotation of the CoG.

You can also use the conversion functions to go from ign-math to the ROS msg, something like:

~~~
  _res->com.position = gazebo_ros::Convert<geometry_msgs::msg::Point>(com.Pos());
  _res->com.orientation = gazebo_ros::Convert<geometry_msgs::msg::Quaternion>(com.Rot());
~~~